pybossa/sched.py

```diff
@@ -300,16 +347,105 @@ def template_get_locked_task(project_id, user_id=None, user_ip=None,
             timeout = timeout or TIMEOUT
             remaining = float('inf') if calibration else n_answers - taskcount
             if acquire_lock(task_id, user_id, remaining, timeout):
+                # reserve tasks
+                acquire_reserve_task_lock(project_id, task_id, user_id, timeout)
                 return _lock_task_for_user(task_id, project_id, user_id, timeout, calibration)
-
         return []
 
     return template_get_locked_task
 
 
+
+def reserve_task_sql_filters(project_id, reserve_task_keys, exclude):
+    # build sql query filter from task category cache key
+    # return sql filter for matching task category keys and list of
+    # task category keys that qualifies for a given project_id
+
+    filters, category_keys = "", []
+
+    if not (project_id and len(reserve_task_keys)):
+        return filters, category_keys
+
+    # convert task category redis cache key to sql query
+    # ex "co:name:IBM:ticker:IBM_US" would be converted to
+    # "task.info->>'co_name' IN ('IBM')"
+    filters_dict = {}
+    for item in reserve_task_keys:
+        regex_key = "reserve_task:project:{}:category:(.+?):user".format(project_id)
+        data = re.search(regex_key, item)
+        if not data:
+            continue
+
+        category_keys += [item]
+        category = data.group(1)
+        category_fv = category.split(":")
+        for i in range(0, len(category_fv), 2):
+            key, value = category_fv[i], category_fv[i + 1]
+            if key in filters_dict:
+                if value not in filters_dict[key]:
+                    filters_dict[key] += [value]
+            else:
+                filters_dict[key] = [value]
+
+    # TODO: pull task # from category keys, look for values from task._add_user_info
+    # generate sql_filter considering value field type instead.
+    data = []
+    for key, value in filters_dict.items():
+        val = ["'{}'".format(val) for val in value]
+        data += ["task.info->>'{}' IN ({})".format(key, ", ".join(val))]
+    if not data:
+        return filters, category_keys
+
+    exclude_clause = "IS NOT TRUE" if exclude else ""
+    filters = "({}) {}".format(" AND ".join(data), exclude_clause)
+    filters = " AND {}".format(filters) if filters else filters
+    return filters, category_keys
+
+
+def get_reserve_task_key(task_id):
+    reserve_key = ""
+    task = task_repo.get_task(task_id)
+    if not task:
+        return reserve_key
+
+    project = project_repo.get(task.project_id)
+    if not (project and project.info.get("sched", "default") in [Schedulers.task_queue]):
+        return reserve_key
+
+    reserve_task_config = project.info.get("reserve_tasks", {}).get("category", [])
+    if not reserve_task_config:
+        return reserve_key
+
+    if not all(field in task.info for field in reserve_task_config):
```

Comment:
Tasks that miss some fields under `reserve_task_config` _key may have non-empty `reserve_key` too, no?

Reply:
`reserve_task_config` contains field names that are part of task reservation category. Hence, if there are tasks that don't have any/some field out of `reserve_task_config`, they'll intended not be reserved.